### DISCUSSION TITLE
Information

### DISCUSSION THREAD

**Discussion by pcpatfire** [pyqtgraph/pyqtgraph#2205] [Q&A] [answered]
I need to draw the trajectories of molecular dynamics simulations (thousands of frames with thousands of atoms each).
Is there an examples of drawing many spheres all together instead of drawing a single sphere? Something like the spheres.py example created by Luke Campagnola with the vispy library.
Thank you

**Comment by j9ac9k** [comment, marked as answer]
Hi @pcpatfire I moved this to the discussion tab as I think this is a better topic for it.

My first thought here is something like the GLScatterPlotItem, there is an example here (the example shows some unrelated functionality too)

https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/examples/GLScatterPlotItem.py

We recently rolled out a very elementary version of a 3D Graph capability; you can see an example here: https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/examples/GLGraphItem.py  The GLGraphItem having the benefit of giving you a way to draw connections between the spheres.

**Comment by pcpatfire** [comment]
Thank you for your answer. The GLScatterPlotItem example is really beautiful!

The following is a picture of the typical situation I have to handle. 
<img width="469" alt="Immagine 2022-02-23 075057" src="https://user-images.githubusercontent.com/50824984/155273275-01b35326-1126-4680-8222-bf3554cf0260.png">
In this case I'm using VisPy and a visual based on the spheres.py code (by Luke). 

There are 'only' 1000 frames with a total number of 3 millions atoms (I can have more than 10000 frames with 30 millions atoms too...), but if I update the plot it takes about 270 seconds! I would like to find a faster method to plot all those atoms. At the beginning I was using the Mayavi library, but it was even slower.

In the picture there is also another problem. The spheres are not respecting the z order of the polyhedra and the box lines added to the plot, which appear behind the spheres even if they are in front of them (is it an Open GL problem?).

Is there an efficient code to draw millions of spheres (possibly with light effects as you can see in the picture)?
Do you think it is possible to do that with pyqtgraph? 

Thank you again
Pacifico Cofrancesco (University of Pavia - Italy)
